### PR TITLE
Improved connection robustness

### PR DESCRIPTION
**Pull request by jonesPD**
By catching many more events that may result in an exception due to a broken connection
By separating intended and non-intended disconnection events such that the library can reconnect automatically

### CODE REVIEW

File: pycomfoconnect/comfoconnect.py [hunk 1]
```diff
@@ -389,16 +443,19 @@ def _message_thread_loop(self):
 
                 elif message.cmd.type == GatewayOperation.GatewayNotificationType:
                     _LOGGER.info('Unhandled GatewayNotificationType')
+                    _LOGGER.debug(message)
```
Comment (michaelarnauts) [inline pycomfoconnect/comfoconnect.py:446]: Why do you want to debug this again? All RX and TX is already logged.

File: pycomfoconnect/comfoconnect.py [hunk 2]
```diff
@@ -351,9 +399,10 @@ def _connect(self, takeover=False):
             self.cmd_register_app(self._local_uuid, self._local_devicename, self._pin, use_queue=False)
 
             # Login
-            self.cmd_start_session(takeover, use_queue=False)
-
-        return True
+            message = self.cmd_start_session(takeover, use_queue=False)
+        
+        return (message)
```
Comment (michaelarnauts) [inline pycomfoconnect/comfoconnect.py:404]: This returns a tuple? Next line should be removed.

**Review by michaelarnauts — CHANGES_REQUESTED** [review]

File: pycomfoconnect/comfoconnect.py [hunk 3]
```diff
@@ -561,8 +618,8 @@ def cmd_rpdo_request(self, pdid: int, type: int = 1, zone: int = 1, timeout=None
     def cmd_keepalive(self, use_queue: bool = True):
         """Sends a keepalive."""
 
-        self._command(
+        reply = self._command(
             KeepAlive,
             use_queue=use_queue
         )
-        return True
+        return (reply == None)
```
Comment (michaelarnauts) [inline pycomfoconnect/comfoconnect.py:625]: No parentheses

File: pycomfoconnect/comfoconnect.py [hunk 4]
```diff
@@ -154,13 +157,26 @@ def connect(self, takeover=False):
...
 
-        # Wait for the background thread to finish
-        self._connection_thread.join()
-        self._connection_thread = None
+        # Wait for the background thread to finish in case it is still active
+        if self._connection_thread != None:
+            self._connection_thread.join()
+            self._connection_thread = None
+
+    def clear_queue(self):
```
Comment (michaelarnauts) [inline pycomfoconnect/comfoconnect.py:169]: Where is this function called?

**Review by michaelarnauts — CHANGES_REQUESTED** [review]

File: pycomfoconnect/comfoconnect.py (hunk 2 above)
Comment (jonesPD) [reply pycomfoconnect/comfoconnect.py:404]: To allow parsing of reply in the calling function(s)

File: pycomfoconnect/comfoconnect.py (hunk 4 above)
Comment (jonesPD) [reply pycomfoconnect/comfoconnect.py:169]: nowhere indeed

File: pycomfoconnect/comfoconnect.py (hunk 1 above)
Comment (jonesPD) [reply pycomfoconnect/comfoconnect.py:446]: They can be removed.

File: pycomfoconnect/comfoconnect.py (hunk 2 above)
Comment (michaelarnauts) [reply pycomfoconnect/comfoconnect.py:404]: Sure, then you just need a return message without parentheses

**Comment by jonesPD** [comment]
I've updated my branch with the changes you requested.

**Review by dkossako — APPROVED** [review]

**Comment by michaelarnauts** [comment]
I'm not convinced that this fixes all issues. The code currently doesn't detect when the bridge is unreachable as long as we don't send a command. I'll need to do something with the keepalives and the tcp timeout to check if the bridge is actually still connected.

**Comment by dkossako** [comment]
> I'm not convinced that this fixes all issues. The code currently doesn't detect when the bridge is unreachable as long as we don't send a command. I'll need to do something with the keepalives and the tcp timeout to check if the bridge is actually still connected.

Okay, I installed library from patch-3 on my side, there are still issues with reconnection, but there are more logs, take a look - maybe that will give you some more details. 

During the day there are some warnings, but all worked fine (so I believe reconnections worked fine here)

> Sep  7 17:13:12 srv hass[1058]: #033[33m2021-09-07 17:13:12 WARNING (Thread-86) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:18:12 srv hass[1058]: #033[33m2021-09-07 17:18:12 WARNING (Thread-87) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:23:13 srv hass[1058]: #033[33m2021-09-07 17:23:13 WARNING (Thread-88) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:28:12 srv hass[1058]: #033[33m2021-09-07 17:28:12 WARNING (Thread-89) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:33:12 srv hass[1058]: #033[33m2021-09-07 17:33:12 WARNING (Thread-90) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:38:12 srv hass[1058]: #033[33m2021-09-07 17:38:12 WARNING (Thread-91) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:43:12 srv hass[1058]: #033[33m2021-09-07 17:43:12 WARNING (Thread-92) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:48:14 srv hass[1058]: #033[33m2021-09-07 17:48:14 WARNING (Thread-93) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:53:12 srv hass[1058]: #033[33m2021-09-07 17:53:12 WARNING (Thread-94) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 17:58:12 srv hass[1058]: #033[33m2021-09-07 17:58:12 WARNING (Thread-95) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 18:03:12 srv hass[1058]: #033[33m2021-09-07 18:03:12 WARNING (Thread-96) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 18:08:12 srv hass[1058]: #033[33m2021-09-07 18:08:12 WARNING (Thread-97) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m
> Sep  7 18:13:12 srv hass[1058]: #033[33m2021-09-07 18:13:12 WARNING (Thread-98) [comfoconnect] THe connection was broken. We will try to reconnect later.#033[0m

Then I had probably some connectivity issues (I got notification on my zabbix regarding network outage) and comfoconnect was not able to connect anymore: https://pastebin.com/VhkUX6rG - HA restart is needed to fix that.

**Comment by hsmade** [comment]
Iirc, the app does do a keepalive command. I remember seeing the same couple of commands every (few) second(s). 
Would it be an idea to implement in this lib as well? Then at least we can actively know if the connection broke and reconnect.

**Comment by michaelarnauts** [comment]
I've moved this logic out of the library in https://github.com/michaelarnauts/aiocomfoconnect. It's up to the user to reconnect if needed, what makes it much simpler.

In Home Assistant, I do it like this:
```python
    @callback
    async def send_keepalive(now) -> None:
        """Send keepalive to the bridge."""
        _LOGGER.debug("Sending keepalive...")
        try:
            await bridge.cmd_keepalive()
        except AioComfoConnectNotConnected:
            # Reconnect when connection has been dropped
            await bridge.connect(entry.data[CONF_LOCAL_UUID])

    entry.async_on_unload(
        async_track_time_interval(hass, send_keepalive, KEEP_ALIVE_INTERVAL)
    )
```

**Pull request closed without merging**